### PR TITLE
change rule of counting gpu available for vc

### PR DESCRIPTION
**Pull request by xudifsd**
count gpu available using new rule

### CODE REVIEW

**Comment by xudifsd** [comment]
Please "Rebase and merge" when merging, I want to preserve my commit history

File: src/docker-images/watchdog/src/watchdog.py [hunk 1]
```diff
@@ -507,69 +518,123 @@ def add_used(self, vc, gpu_type, count):
...
+        for vc_name, gpu_info in ratio.items():
+            for gpu_type, cur_ratio in gpu_info.items():
+                ratio_sum += cur_ratio
+
+        for vc_name, gpu_info in ratio.items():
+            for gpu_type, cur_ratio in gpu_info.items():
+                if vc_name not in vc_usage.map or gpu_type not in vc_usage.map[vc_name]:
+                    labels = [vc_name, gpu_type]
+                    # no job running in this vc or using this gpu type
+                    available = math.floor(cluster_gpu_info.available * cur_ratio / ratio_sum)
```
Comment (Anbang-Hu) [inline src/docker-images/watchdog/src/watchdog.py]: Cast to `int`

File: src/docker-images/watchdog/src/watchdog.py [hunk 2]
```diff
@@ -507,69 +518,123 @@ def add_used(self, vc, gpu_type, count):
...
+
+                if gpu_type not in vc_info[vc_name]:
+                    logger.warning("ignore used gpu %s in %s, but vc quota do not have this gpu_type", gpu_type, vc_name)
+                    continue
+
+                labels = [vc_name, gpu_type]
+
+                cur_ratio = ratio[vc_name][gpu_type]
+                quota = vc_info[vc_name][gpu_type]
+                available = math.floor(cluster_gpu_info.available * cur_ratio / ratio_sum)
```
Comment (Anbang-Hu) [inline src/docker-images/watchdog/src/watchdog.py]: Cast to `int`

File: src/docker-images/watchdog/src/watchdog.py [hunk 3]
```diff
@@ -507,69 +518,123 @@ def add_used(self, vc, gpu_type, count):
...
+
+        gpu_unallocatable = cluster_gpu_info.capacity - cluster_gpu_info.allocatable
+
+        # key is vc_name, value is a map with key to be gpu_type and value to be real
+        # quota
+        ratio = collections.defaultdict(lambda : {})
+
+        for vc_name, gpu_info in vc_info.items():
+            for gpu_type, quota in gpu_info.items():
+                vc_quota = quota - math.ceil(gpu_unallocatable * quota / vc_quota_sum)
```
Comment (Anbang-Hu) [inline src/docker-images/watchdog/src/watchdog.py]: Cast to `int`

File: src/docker-images/watchdog/src/watchdog.py [hunk 4]
```diff
@@ -507,69 +518,123 @@ def add_used(self, vc, gpu_type, count):
...
+
+        gpu_unallocatable = cluster_gpu_info.capacity - cluster_gpu_info.allocatable
+
+        # key is vc_name, value is a map with key to be gpu_type and value to be real
+        # quota
+        ratio = collections.defaultdict(lambda : {})
+
+        for vc_name, gpu_info in vc_info.items():
+            for gpu_type, quota in gpu_info.items():
+                vc_quota = int(quota - math.ceil(gpu_unallocatable * quota / vc_quota_sum))
```
Comment (Anbang-Hu) [inline src/docker-images/watchdog/src/watchdog.py]: Corner case `vc_quota_sum == 0`

File: src/docker-images/watchdog/src/watchdog.py [hunk 5]
```diff
@@ -507,69 +518,123 @@ def add_used(self, vc, gpu_type, count):
...
+        for vc_name, gpu_info in ratio.items():
+            for gpu_type, cur_ratio in gpu_info.items():
+                ratio_sum += cur_ratio
+
+        for vc_name, gpu_info in ratio.items():
+            for gpu_type, cur_ratio in gpu_info.items():
+                if vc_name not in vc_usage.map or gpu_type not in vc_usage.map[vc_name]:
+                    labels = [vc_name, gpu_type]
+                    # no job running in this vc or using this gpu type
+                    available = int(math.floor(cluster_gpu_info.available * cur_ratio / ratio_sum))
```
Comment (Anbang-Hu) [inline src/docker-images/watchdog/src/watchdog.py]: Corner case `ratio_sum == 0`

File: src/docker-images/watchdog/src/watchdog.py [hunk 6]
```diff
@@ -507,69 +518,123 @@ def add_used(self, vc, gpu_type, count):
...
+
+                if gpu_type not in vc_info[vc_name]:
+                    logger.warning("ignore used gpu %s in %s, but vc quota do not have this gpu_type", gpu_type, vc_name)
+                    continue
+
+                labels = [vc_name, gpu_type]
+
+                cur_ratio = ratio[vc_name][gpu_type]
+                quota = vc_info[vc_name][gpu_type]
+                available = int(math.floor(cluster_gpu_info.available * cur_ratio / ratio_sum))
```
Comment (Anbang-Hu) [inline src/docker-images/watchdog/src/watchdog.py]: Corner case `ratio_sum == 0`

**Pull request merged**